### ISSUE TITLE
"Buffer too small, skipping audio rendering"

### ISSUE THREAD

**Issue by hejops** [agourlay/ruxguitar#14]
https://github.com/agourlay/ruxguitar/blob/21de214cc7b22944e1545ad5d13272d3282d7467/src/audio/midi_player.rs#L308-L311

I added some logging statements, and found out that `left.len()` and `right.len()` are 2205, while `channel_len` is 4704 -- this is true for any `.gp5` file (like the `Demo v5.gp5`) and any soundfont. I don't know much about MIDI rendering, so what are the typical values for each length? What information can I provide to better help debug this error?

I'm running on Arch Linux, with [`alsa-lib`](https://archlinux.org/packages/extra/x86_64/alsa-lib/) installed (which provides `/usr/lib/libasound.so`).

Thanks for undertaking this project, by the way; I've always wanted to have a go at writing a GP5 player but never found the time/motivation!

**Comment by agourlay** [comment]
Thank you for the report and the kind words :+1: 

Can you please try this branch https://github.com/agourlay/ruxguitar/pull/15 to investigate this issue.

I would like to know how it sounds and what is logged :)

**Comment by agourlay** [comment]
@hejops ping just in case you missed the notification :)

**Comment by hejops** [comment]
My bad, I've been a bit swamped with other work lately; I'll look into this tomorrow!

**Comment by hejops** [comment]
Here are the logs from playing `Demo v5.gp5`:

```sh
$ cargo run -- --file test-files/Demo\ v5.gp5
[2024-09-19T05:17:59Z INFO  ruxguitar::audio::midi_player] Toggle audio stream
[2024-09-19T05:18:00Z WARN  ruxguitar::audio::midi_player] Audio output stream config: StreamConfig { channels: 2, sample_rate: SampleRate(44100), buffer_size: Default } (2 channels, 44100 Hz)
[2024-09-19T05:18:00Z INFO  ruxguitar::audio::midi_player] Buffer too small 2205<4704
[2024-09-19T05:18:54Z INFO  ruxguitar::audio::midi_player] Tempo changed to 120
[2024-09-19T05:19:02Z INFO  ruxguitar::audio::midi_player] Tempo changed to 165
```

(PS: See https://github.com/hejops/ruxguitar/commit/9b2d470f156114fc60810efaa8309db30d9327ec for how/why I added the `--file` cli argument)

Could the lack of output have something to do with `alsa` and/or `pulseaudio`? It seems that `ruxguitar` output is not being detected in `pulseaudio`; for example, playing audio in Firefox or mpv can typically be observed through `pactl` by using something like:

```sh
$ pactl list | grep module-stream-restore.id
                module-stream-restore.id = "sink-input-by-application-name:Firefox"
                module-stream-restore.id = "sink-input-by-application-name:mpv Media Player"
```

But playing a file in `ruxguitar` produces no output.

```sh
$ pactl list | grep ruxguitar
```

I haven't had to fiddle with `alsa` in a long time (or `pulseaudio` for that matter), so maybe I'm missing something obvious.

**Comment by agourlay** [comment]
Thanks for sharing the logs :+1: 

It seems the output stream is properly created so it is very weird that you have no sounds indeed.

FYI I am using Pipewire and Alsa and when running: 

```
$ pactl list | grep ruxguitar
```

I get:

```
Client #721
        Driver: PipeWire
        Owner Module: 2
        Properties:
                pipewire.protocol = "protocol-native"
                pipewire.sec.socket = "pipewire-0"
                pipewire.sec.pid = "3436"
                pipewire.sec.uid = "1000"
                pipewire.sec.gid = "1000"
                pipewire.sec.label = "unconfined"
                module.id = "2"
                object.id = "79"
                object.serial = "721"
                client.api = "pipewire-pulse"
                pulse.server.type = "unix"
                application.name = "ALSA plug-in [ruxguitar]"
                application.process.id = "211737"
                application.process.user = "agourlay"
                application.process.host = "agourlay-thinkpad"
                application.process.binary = "ruxguitar"
                application.language = "C"
                window.x11.display = ":0.0"
                application.process.machine_id = "374a11792ee84724a58ed2202bbb6619"
                application.process.session_id = "c2"
                config.name = "pipewire-pulse.conf"
                cpu.max-align = "32"
                default.clock.rate = "48000"
                default.clock.quantum = "1024"
                default.clock.min-quantum = "32"
                default.clock.max-quantum = "2048"
                default.clock.quantum-limit = "8192"
                default.clock.quantum-floor = "4"
                default.video.width = "640"
                default.video.height = "480"
                default.video.rate.num = "25"
                default.video.rate.denom = "1"
                log.level = "2"
                clock.power-of-two-quantum = "true"
                link.max-buffers = "64"
                mem.warn-mlock = "false"
                mem.allow-mlock = "true"
                settings.check-quantum = "false"
                settings.check-rate = "false"
                core.version = "1.0.5"
                core.name = "pipewire-agourlay-3436"
                pipewire.access = "unrestricted"
```

However, looking at the docs for the [cpal](https://github.com/RustAudio/cpal) crate that I am using, it does not seem to mention supporting PulseAudio though.

```
Currently, supported hosts include:

    Linux (via ALSA or JACK)
``` 

See tracking issue https://github.com/RustAudio/cpal/issues/259

Is it possible for you to use Pipewire?

**Comment by hejops** [comment]
Thanks for investigating. On replacing `pulseaudio` with `pipewire-pulse`, `cpal` panics at `device.default_output_config()` because it cannot find any device:

```text
[2024-09-20T04:35:41Z INFO  ruxguitar::audio::midi_player] Toggle audio stream
ALSA lib pcm_dmix.c:1000:(snd_pcm_dmix_open) unable to open slave
thread 'main' panicked at src/audio/midi_player.rs:213:49:
called `Result::unwrap()` on an `Err` value: DeviceNotAvailable
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I've added submitted a PR to avoid panicking in such a situation, feel free to merge it if you think it looks good. :smiley: 

I'll be sure to give it another try when I get some better speakers in the (hopefully) near future!

**Comment by agourlay** [comment]
fixed in [0.3.1](https://github.com/agourlay/ruxguitar/releases/tag/v0.3.1)
